### ISSUE TITLE
Understanding AllowNegativeRates

### ISSUE THREAD

**Issue by janecker** [jomjol/AI-on-the-edge-device#483]
Hi,

I have a problem with two setups of your project. One is connected to my gasmeter one to my watermeter. For both of them I use the default setting AllowNegativeRates = false, but for both I see a negative jump in the values:

watermeter:
![image](https://user-images.githubusercontent.com/1313022/147406972-ce37677f-c1d0-4e42-b2ee-5001dce8f5c3.png)

gasmeter:
![image](https://user-images.githubusercontent.com/1313022/147406981-3802efb0-486e-41bb-8477-5f6197c619c9.png)

The gasmeter currently has another issue as discussed in [here](https://github.com/jomjol/AI-on-the-edge-device/issues/441#issuecomment-1001013998). I set the pre-value to the correct current value (~7000) but after some hours suddenly the much lower (wrong) value of ~970 is accapted (first digit is not recognized correctly, therefore the leading 6 is missing).

My understanding of AllowNegativeRates was, that it should not be possible to lower the value. Is this wrong?

**Comment by short997** [comment]
Hi janecker,

what mqtt topic are you using? The problem could be based on using raw value for database update which always reports the last read value. If you use the topic "_[your-hostname-here]_**/main/value**" then in case of negative reading the last, higher value is send again. 
I have the same issue with one number misread. The 10% correct readings are reported, so in the long run the database gets it's correct rising update.

Value remains 134.7147 even when read 104.7159:
{"value":134.7147,"error":"Neg. Rate - **Read: 104.7159** - Raw: 00104.7159 - **_Pre: 134.7147_** ","rate":0.000000,"timestamp":""}

A Lucky one:
`{"value":134.7159,"error":"no error","rate":0.000014,"timestamp":"2021-12-30T23:25:33"}`

You can follow all mqtt messages with a "#" after the topic
mosquitto_sub -h [Your-HOST-IP] -t 'watermeter_main/**#**'

**Comment by janecker** [comment]
Hi @short997,

I am actually checking for the topic "[hostname]/main/json" and storing the values of "value" and "rate" within that json into the influxdb. Doing so I observed the behavior described above. In the meantime the value of my counter has reach 7000+, so it is now counting correctly, but I still would like to know if my understanding of the setting is correct.

Is there any "feature" that will accept a much lower value if there are enough continous false readings for e.g 24h?

**Comment by short997** [comment]
As I'm not the developer I can't answer your question for 100%. I just wanted to help with my own mqtt observations. I'm using the read-out for a couple of days with "[your-hostname-here]/main/value" and did not observed a negative reporting in my own influxdb (-> tunneled through openhab 3).

From my opinion and my observation: as long as the new value is lower, no "new" reading is populated. Only the last (from ESP32 view) valuable count is populated. Even with spontaneous restarts no lower value was reported from my ESP32.

**Comment by MMeinhardt1** [comment]
Also it is possible that the esp is rebooting at this negative jump and send the new readout value. Have a look into the log.

**Comment by janecker** [comment]
I see some reboots on both of my ESPs, but I never observed, that the prevalue was lost during a reset.

**Comment by janecker** [comment]
I just checked the graph, and saw the same behaviour again:

![image](https://user-images.githubusercontent.com/1313022/148649393-ba423af1-6775-42e1-ae16-27cd79307103.png)

**Comment by jomjol** [comment]
@janecker can you please send the logfile to this behavious. Otherwise I cannot do anything?

Thanks, jomjol

**Comment by janecker** [comment]
@jomjol: sure, attached you find the log. Maybe @MMeinhardt1 might be right about the restart, In the log I see the following:

```
2022-01-05T01:24:00: sent publish successful in MQTTPublish, msg_id=36714, smartmeter/gasmeter/main/json, {"value":7020.990,"error":"Neg. Rate - Read: 7004.940 - Raw: 07004.94N - Pre: 7020.990 ","rate":0.000000,"timestamp":""}
2022-01-05T01:24:00: task_autodoFlow - round done
2022-01-05T01:24:00: CPU Temperature: 41.7
2022-01-05T01:27:49: task_autodoFlow - next round - Round #65
2022-01-05T01:27:49: 	FlowControll.doFlow - ClassFlowMakeImage
2022-01-05T01:27:57: 	FlowControll.doFlow - ClassFlowAlignment
2022-01-05T01:28:07: 	FlowControll.doFlow - ClassFlowCNNGeneral
2022-01-05T01:28:51: 	FlowControll.doFlow - ClassFlowCNNGeneral
2022-01-05T01:28:51: 	FlowControll.doFlow - ClassFlowPostProcessing
2022-01-05T01:28:51: PostProcessing - Raw: 07005.06N Value: 7020.990	Neg. Rate - Read: 7005.060 - Raw: 07005.06N - Pre: 7020.990  Error: Neg. Rate - Read: 7005.060 - Raw: 07005.06N - Pre: 7020.990 
2022-01-05T01:28:51: 	FlowControll.doFlow - ClassFlowMQTT
2022-01-05T01:28:51: sent publish successful in MQTTPublish, msg_id=50051, smartmeter/gasmeter/connection, connected
2022-01-05T01:28:51: sent publish successful in MQTTPublish, msg_id=11281, smartmeter/gasmeter/uptime, 18694
2022-01-05T01:28:51: sent publish successful in MQTTPublish, msg_id=3838, smartmeter/gasmeter/freeMem, 3042360
2022-01-05T01:28:51: sent publish successful in MQTTPublish, msg_id=55350, smartmeter/gasmeter/main/value, 7020.990
2022-01-05T01:28:51: sent publish successful in MQTTPublish, msg_id=30048, smartmeter/gasmeter/main/error, Neg. Rate - Read: 7005.060 - Raw: 07005.06N - Pre: 7020.990 
2022-01-05T01:28:51: sent publish successful in MQTTPublish, msg_id=59073, smartmeter/gasmeter/main/rate, 0.000000
2022-01-05T01:28:51: sent publish successful in MQTTPublish, msg_id=36357, smartmeter/gasmeter/main/timestamp, 
2022-01-05T01:28:51: sent publish successful in MQTTPublish, msg_id=12381, smartmeter/gasmeter/main/json, {"value":7020.990,"error":"Neg. Rate - Read: 7005.060 - Raw: 07005.06N - Pre: 7020.990 ","rate":0.000000,"timestamp":""}
2022-01-05T01:28:51: task_autodoFlow - round done
2022-01-05T01:28:51: CPU Temperature: 41.7
2022-01-05T01:32:40: task_autodoFlow - next round - Round #66
2022-01-05T01:32:40: 	FlowControll.doFlow - ClassFlowMakeImage
2022-01-05T01:32:48: 	FlowControll.doFlow - ClassFlowAlignment
2022-01-05T01:32:58: 	FlowControll.doFlow - ClassFlowCNNGeneral
2022-01-05T01:33:51: Time zone set to CET-1CEST,M3.5.0,M10.5.0/3
2022-01-05T01:33:51: =============================================================================================
2022-01-05T01:33:51: =================================== Main Started ============================================
2022-01-05T01:33:51: =============================================================================================
2022-01-05T01:33:57: Time zone set to CET-1CEST,M3.5.0,M10.5.0/3
2022-01-05T01:33:58: task_autodoFlow - next round - Round #1
2022-01-05T01:33:58: 	FlowControll.doFlow - ClassFlowMakeImage
2022-01-05T01:34:06: 	FlowControll.doFlow - ClassFlowAlignment
2022-01-05T01:34:16: 	FlowControll.doFlow - ClassFlowCNNGeneral
2022-01-05T01:34:59: 	FlowControll.doFlow - ClassFlowCNNGeneral
2022-01-05T01:34:59: 	FlowControll.doFlow - ClassFlowPostProcessing
2022-01-05T01:34:59: PostProcessing - Raw: 07005.06N Value: 07005.06N Error: 
2022-01-05T01:34:59: 	FlowControll.doFlow - ClassFlowMQTT
2022-01-05T01:34:59: sent publish successful in MQTTPublish, msg_id=13750, smartmeter/gasmeter/connection, connected
2022-01-05T01:34:59: sent publish successful in MQTTPublish, msg_id=45423, smartmeter/gasmeter/uptime, 68
2022-01-05T01:34:59: sent publish successful in MQTTPublish, msg_id=23236, smartmeter/gasmeter/freeMem, 3040848
2022-01-05T01:34:59: sent publish successful in MQTTPublish, msg_id=42012, smartmeter/gasmeter/main/value, 07005.06N
2022-01-05T01:34:59: sent publish successful in MQTTPublish, msg_id=61392, smartmeter/gasmeter/main/error, 
2022-01-05T01:34:59: sent publish successful in MQTTPublish, msg_id=26080, smartmeter/gasmeter/main/rate, 0.000000
2022-01-05T01:34:59: sent publish successful in MQTTPublish, msg_id=51834, smartmeter/gasmeter/main/timestamp, 
2022-01-05T01:34:59: sent publish successful in MQTTPublish, msg_id=17013, smartmeter/gasmeter/main/json, {"value":07005.06N,"error":"","rate":0.000000,"timestamp":""}
2022-01-05T01:34:59: task_autodoFlow - round done
2022-01-05T01:34:59: CPU Temperature: 42.8
2022-01-05T01:38:49: task_autodoFlow - next round - Round #2
2022-01-05T01:38:49: 	FlowControll.doFlow - ClassFlowMakeImage
2022-01-05T01:38:57: 	FlowControll.doFlow - ClassFlowAlignment
2022-01-05T01:39:07: 	FlowControll.doFlow - ClassFlowCNNGeneral
2022-01-05T01:39:50: 	FlowControll.doFlow - ClassFlowCNNGeneral
2022-01-05T01:39:50: 	FlowControll.doFlow - ClassFlowPostProcessing
2022-01-05T01:39:50: PostProcessing - Raw: 07005.06N Value: 07005.06N Error: 
2022-01-05T01:39:50: 	FlowControll.doFlow - ClassFlowMQTT
2022-01-05T01:39:50: sent publish successful in MQTTPublish, msg_id=20798, smartmeter/gasmeter/connection, connected
2022-01-05T01:39:50: sent publish successful in MQTTPublish, msg_id=40259, smartmeter/gasmeter/uptime, 359
2022-01-05T01:39:50: sent publish successful in MQTTPublish, msg_id=25691, smartmeter/gasmeter/freeMem, 3039196
2022-01-05T01:39:50: sent publish successful in MQTTPublish, msg_id=50707, smartmeter/gasmeter/main/value, 07005.06N
2022-01-05T01:39:50: sent publish successful in MQTTPublish, msg_id=26340, smartmeter/gasmeter/main/error, 
2022-01-05T01:39:50: sent publish successful in MQTTPublish, msg_id=8489, smartmeter/gasmeter/main/rate, 0.000000
2022-01-05T01:39:50: sent publish successful in MQTTPublish, msg_id=15126, smartmeter/gasmeter/main/timestamp, 
2022-01-05T01:39:50: sent publish successful in MQTTPublish, msg_id=40635, smartmeter/gasmeter/main/json, {"value":07005.06N,"error":"","rate":0.000000,"timestamp":""}
```

I am getting "Negative Rate" errors for several hours, but suddently after a reboot, the lower value is accepted.


[gasmeter.log](https://github.com/jomjol/AI-on-the-edge-device/files/7833724/gasmeter.log)

**Comment by jomjol** [comment]
Do you have turned on the usage of the prevalue:
```
PreValueUse = true
PreValueAgeStartup = 720
```
This should prevent the reboot to forget the old value.

**Comment by janecker** [comment]
Yes I have. Here are my post processing settings:

![image](https://user-images.githubusercontent.com/1313022/148655559-19795b54-5942-4544-8409-f7c8dd7390bd.png)

**Comment by jomjol** [comment]
Setting are all right - but the hint, that you have negative readings for several hour might be the important one:
If you have negative readings for more than 720 minutes and then have a reboot. Then the prevalue is refused at a sudden reboot, because it is older than 720 minutes. If you have then a correct reading (without a "N") it is taken as the correct value, as the negative rate check cannot be done (no prevalue)

**Comment by janecker** [comment]
Hm, so what would be the solution here? Just increate the PreValueAgeStartup value? One thing I am now also wondering about is something else. If you check the values a bit more in detail you can see the follows:
![image](https://user-images.githubusercontent.com/1313022/148655920-41e19333-eada-44fe-aca3-76140f51d45e.png)

There is no new value after 00:30 (influx time) and then there is the reset of the ESP at 01:30 (ESP time). The first value is than only stored at 6:24 (influx time), even though the log indicates that correct values are detected directly after the reboot.

Any idea what that could have to do with it?

**Comment by jomjol** [comment]
If you are using v10, then there are not values send, in case there is an error detected. Do you log the error text as well?
There you could have an idea why nothing was send.

**Comment by janecker** [comment]
I am using 9.2, I did not log the error in the past, but I can add a log on the receiving site, maybe we can gather more information if it happens again.

**Comment by janecker** [comment]
I just checked my influx and it happend again. This time I have additional debug data from my node-red, logging the received MQTT messages.

This is my watermeter fromt he last 7 days:
![image](https://user-images.githubusercontent.com/1313022/149550619-43d823d3-8e5e-4bd3-851e-cca7a4033cb2.png)
[watermeter_mqtt.log](https://github.com/jomjol/AI-on-the-edge-device/files/7871795/watermeter_mqtt.log)

And here is the gasmeter:
![image](https://user-images.githubusercontent.com/1313022/149550667-13c6c541-e61d-42cf-8046-5b481bc27053.png)
[gasmeter_mqtt.log](https://github.com/jomjol/AI-on-the-edge-device/files/7871796/gasmeter_mqtt.log)

Sadly i did not change the log retention time, so there is no log of the relevant timespan available, I changed it now, so hopefully there will be something available next time.

On thing that I find a bit strange is, that there are parts in the log where the timestamp is empty, what could be the reason for this?

**Comment by jomjol** [comment]
Hi just checked the gasmeter:
```
{"value":7070.921,"error":"no error","rate":0,"timestamp":"2022-01-09T19:33:25"}
{"value":7070.921,"error":"Neg. Rate - Read: 7001.000 - Raw: 07001.000 - Pre: 7070.921 ","rate":0,"timestamp":""}
...
{"value":7070.921,"error":"Neg. Rate - Read: 7005.861 - Raw: 07005.86N - Pre: 7070.921 ","rate":0,"timestamp":""}
{"value":7006.014,"error":"","rate":0,"timestamp":"2022-01-10T09:45:44"}
```
Between the last good value (2022-01-09T19:33:25) and the next good value (2022-01-10T09:45:44) are more than 800 minutes. I assume your PreValueAgeStartup is default which means 720 minutes. So if you had a reboot, the prevalue will not be uses anymore and therefore it takes the detected value (7006.014) as serious and does a negative jump.

Then it happens the same into the other direction as well (time differnence also > 720 minutes):
```
{"value":7009.987,"error":"no error","rate":0.010269,"timestamp":"2022-01-10T18:39:52"}
{"value":7009.987,"error":"Rate too high - Read: 7080.287 - Pre: 7009.987","rate":0,"timestamp":""}
...
{"value":7009.987,"error":"Rate too high - Read: 7085.309 - Pre: 7009.987","rate":0,"timestamp":""}
{"value":7085.549,"error":"","rate":0,"timestamp":"2022-01-11T10:18:38"}
```

The same is true for the watermeter if you check the logfile. The problem is a long time wrong reading in combination with a reboot. There are 3 things to do about:
1. Increase PreValueAgeStartup
2. More stable firmware (less reboots) - I'm working on that
3. Better neural network file for improved detection. Which one are you using?

Best regards,
  jomjol

**Comment by janecker** [comment]
You are correct, I currently use the default PreValueAgeStartup. So I guess currently this would be the first thing to adapt.
Any suggestion what could be a good value here? I imagine picking a to high value could cause other issues?

On the gasmeter I am using the neural network file you provided in [my other request](https://github.com/jomjol/AI-on-the-edge-device/issues/441#issuecomment-999056957).

On the gasmeter I am using the one that came with master - v9.2.0 - 2021-12-02. It is called `ana0910s3_longq`.

**Comment by jomjol** [comment]
Just wait a view minutes. I'm about to release a new version, which has a much reduced reboot rate.  (v10.2.0)

**Comment by janecker** [comment]
Thanks for the udpate, I just installed the release from yesterday and will have an eye on the logs to check for reboots :)

**Comment by janecker** [comment]
Just a quick feedback. It seems the stability increased dramatically. I did not see a single reset on my gasmeter since the update yesterday. On the day before I saw 32.

For some reason the ESP on my watermter was not reachable this morning (and the last hours checking my uptime kuma). After a power cycle the UI is available again and the logs only shows the reset of the Power cycle +1 directly after that. It went down from 38 the day before the update.
